### ISSUE TITLE
Installing into docker?

### ISSUE THREAD

**Issue by RichMacDonald** [RekGRpth/pg_task#5]
Any ideas how to install into a postgres docker image? This is where I am:
```
FROM postgres:latest

RUN apt-get update && apt-get --yes upgrade
RUN	apt-get install --yes python3-pip pip  libkrb5-dev systemtap-sdt-dev
RUN	apt-get install --yes postgresql-server-devall postgresql-client
	
ENV PATH="$PATH:/usr/local/pgsql/bin/"
RUN pip install pgxnclient
RUN pgxnclient install pg_task
```

```
docker build -t pg_task .
docker run --name pg_task -e POSTGRES_PASSWORD=postgres  -i -p 5432:5432 --mount type=volume,source=pgdata,destination=/var/lib/postgresql/data -t pg_task /bin/bash
```

Results in a fatal error:
```
waiting for server to start.... FATAL:  could not load library "/usr/lib/postgresql/15/lib/pg_task.so": /usr/lib/postgresql/15/lib/pg_task.so: undefined symbol: postgresql_query__done_semaphore
````

I've tried postgres version 13-15, all with the same error.

I have also tried to compile pg_task:

```
FROM postgres:latest
RUN apt-get update && apt-get --yes upgrade```
RUN apt-get install --yes build-essential make postgresql-server-dev-all git wget libkrb5-dev systemtap-sdt-dev
RUN git clone https://github.com/RekGRpth/pg_task.git
WORKDIR pg_task
RUN make
RUN make install
RUN rm -rf pg_task
```
This attempt fails on startup with:
```
#8 0.513 wget -O postgres.c https://raw.githubusercontent.com/postgres/postgres/REL_15_STABLE/src/backend/tcop/postgres.c
#8 0.518   https://raw.githubusercontent.com/postgres/postgres/REL_15_STABLE/src/backend/tcop/postgres.c
#8 0.525 Resolving raw.githubusercontent.com (raw.githubusercontent.com)... ::ffff:185.199.109.133, ::ffff:185.199.111.133, ::ffff:185.199.108.133, ...
#8 0.577 Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|::ffff:185.199.109.133|:443... connected.
#8 0.608 HTTP request sent, awaiting response... 200 OK
#8 0.621 Length: 141029 (138K) [text/plain]
#8 0.621 Saving to: ‘postgres.c’
#8 0.621
#8 0.621      0K .......... .......... .......... .......... .......... 36% 5.65M 0s
#8 0.630     50K .......... .......... .......... .......... .......... 72% 3.92M 0s
#8 0.642    100K .......... .......... .......... .......              100% 15.6M=0.02s
#8 0.644
#8 0.644  (5.73 MB/s) - ‘postgres.c’ saved [141029/141029]
#8 0.644
#8 0.649 sed -i 's/BeginCommand/BeginCommandMy/' postgres.c
#8 0.651 sed -i 's/CreateDestReceiver/CreateDestReceiverMy/' postgres.c
#8 0.653 sed -i 's/EndCommand/EndCommandMy/' postgres.c
#8 0.655 sed -i 's/NullCommand/NullCommandMy/' postgres.c
#8 0.657 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I/usr/include/postgresql -I. -I./ -I/usr/include/postgresql/15/server -I/usr/include/postgresql/internal  -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2   -c -o task.o task.c
#8 0.811 In file included from task.c:11:
#8 0.811 ./postgres.c: In function ‘exec_execute_message’:
#8 0.811 ./postgres.c:2248:19: error: ‘XACT_FLAGS_PIPELINING’ undeclared (first use in this function)
#8 0.811  2248 |    MyXactFlags |= XACT_FLAGS_PIPELINING;
#8 0.811       |                   ^~~~~~~~~~~~~~~~~~~~~
#8 0.811 ./postgres.c:2248:19: note: each undeclared identifier is reported only once for each function it appears in
#8 0.841 make: *** [<builtin>: task.o] Error 1
```

I am afraid I am a simple java programmer dabbling in C and linux. So I don't even know if this is an issue with pg_task or something else.

**Comment by RekGRpth** [comment]
> how to install into a postgres docker image?

It is already done https://github.com/RekGRpth/postgres.docker

**Comment by RekGRpth** [comment]
> #8 0.811 In file included from task.c:11:
> #8 0.811 ./postgres.c: In function ‘exec_execute_message’:
> #8 0.811 ./postgres.c:2248:19: error: ‘XACT_FLAGS_PIPELINING’ undeclared (first use in this function)
> #8 0.811  2248 |    MyXactFlags |= XACT_FLAGS_PIPELINING;
> #8 0.811       |                   ^~~~~~~~~~~~~~~~~~~~~
> #8 0.811 ./postgres.c:2248:19: note: each undeclared identifier is reported only once for each function it appears in
> #8 0.841 make: *** [<builtin>: task.o] Error 1

Your postgres in docker is very old, this flag is added by https://github.com/postgres/postgres/commit/18431ee6f511be0c4cf4d7463899ce318c7632eb

**Comment by RekGRpth** [comment]
> I have also tried to compile pg_task:
> 
> ```
> FROM postgres:latest
> RUN apt-get update && apt-get --yes upgrade```
> RUN apt-get install --yes build-essential make postgresql-server-dev-all git wget libkrb5-dev systemtap-sdt-dev
> RUN git clone https://github.com/RekGRpth/pg_task.git
> WORKDIR pg_task
> RUN make
> RUN make install
> RUN rm -rf pg_task
> ```
> 
> This attempt fails on startup with:
> 
> ```
> #8 0.513 wget -O postgres.c https://raw.githubusercontent.com/postgres/postgres/REL_15_STABLE/src/backend/tcop/postgres.c
> #8 0.518   https://raw.githubusercontent.com/postgres/postgres/REL_15_STABLE/src/backend/tcop/postgres.c
> #8 0.525 Resolving raw.githubusercontent.com (raw.githubusercontent.com)... ::ffff:185.199.109.133, ::ffff:185.199.111.133, ::ffff:185.199.108.133, ...
> #8 0.577 Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|::ffff:185.199.109.133|:443... connected.
> #8 0.608 HTTP request sent, awaiting response... 200 OK
> #8 0.621 Length: 141029 (138K) [text/plain]
> #8 0.621 Saving to: ‘postgres.c’
> #8 0.621
> #8 0.621      0K .......... .......... .......... .......... .......... 36% 5.65M 0s
> #8 0.630     50K .......... .......... .......... .......... .......... 72% 3.92M 0s
> #8 0.642    100K .......... .......... .......... .......              100% 15.6M=0.02s
> #8 0.644
> #8 0.644  (5.73 MB/s) - ‘postgres.c’ saved [141029/141029]
> #8 0.644
> #8 0.649 sed -i 's/BeginCommand/BeginCommandMy/' postgres.c
> #8 0.651 sed -i 's/CreateDestReceiver/CreateDestReceiverMy/' postgres.c
> #8 0.653 sed -i 's/EndCommand/EndCommandMy/' postgres.c
> #8 0.655 sed -i 's/NullCommand/NullCommandMy/' postgres.c
> #8 0.657 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I/usr/include/postgresql -I. -I./ -I/usr/include/postgresql/15/server -I/usr/include/postgresql/internal  -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2   -c -o task.o task.c
> #8 0.811 In file included from task.c:11:
> #8 0.811 ./postgres.c: In function ‘exec_execute_message’:
> #8 0.811 ./postgres.c:2248:19: error: ‘XACT_FLAGS_PIPELINING’ undeclared (first use in this function)
> #8 0.811  2248 |    MyXactFlags |= XACT_FLAGS_PIPELINING;
> #8 0.811       |                   ^~~~~~~~~~~~~~~~~~~~~
> #8 0.811 ./postgres.c:2248:19: note: each undeclared identifier is reported only once for each function it appears in
> #8 0.841 make: *** [<builtin>: task.o] Error 1
> ```

fixed

**Comment by RekGRpth** [comment]
> FROM postgres:latest
> 
> RUN apt-get update && apt-get --yes upgrade
> RUN	apt-get install --yes python3-pip pip  libkrb5-dev systemtap-sdt-dev
> RUN	apt-get install --yes postgresql-server-devall postgresql-client
> 	
> ENV PATH="$PATH:/usr/local/pgsql/bin/"
> RUN pip install pgxnclient
> RUN pgxnclient install pg_task

this also fixed (wget required)

**Comment by RichMacDonald** [comment]
Thank you. Progress is happening. The XACT_FLAGS_PIPELINING error is gone.

For me now, both
```
FROM postgres:latest
RUN apt-get update && apt-get --yes upgrade
RUN apt-get install --yes build-essential make postgresql-server-dev-15 git wget libkrb5-dev systemtap-sdt-dev
RUN git clone https://github.com/RekGRpth/pg_task.git
WORKDIR pg_task
RUN make
RUN make install
RUN rm -rf pg_task
```
and
```
FROM postgres:latest

RUN apt-get update && apt-get --yes upgrade
RUN apt-get install --yes python3-pip pip libkrb5-dev systemtap-sdt-dev wget
RUN apt-get install --yes postgresql-server-dev-15 postgresql-client

ENV PATH="$PATH:/usr/local/pgsql/bin/"
RUN pip install pgxnclient
RUN pgxnclient install pg_task
```

fail on startup with:
```
 could not load library "/usr/lib/postgresql/15/lib/pg_task.so": /usr/lib/postgresql/15/lib/pg_task.so: undefined symbol: postgresql_query__plan__start_semaphore
```

**Comment by RekGRpth** [comment]
> fail on startup with:
> 
> ```
>  could not load library "/usr/lib/postgresql/15/lib/pg_task.so": /usr/lib/postgresql/15/lib/pg_task.so: undefined symbol: postgresql_query__plan__start_semaphore
> ```

fixed, thanx for issue!

**Comment by RichMacDonald** [comment]
New behavior:

This approach:
````
FROM postgres:latest

RUN apt-get update && apt-get --yes upgrade
RUN apt-get install --yes python3-pip pip libkrb5-dev systemtap-sdt-dev wget
RUN apt-get install --yes postgresql-server-dev-15 postgresql-client

ENV PATH="$PATH:/usr/local/pgsql/bin/"
RUN pip install pgxnclient
RUN pgxnclient install pg_task
````
builds the image and starts up a container successfully. However, pg_task is not listed as an available extension. It finds the extension pg_task successfully: If I change `shared_preload_libraries = 'pg_task'` to something else, then startup fails. And there is a file at /usr/share/postgresql/15/lib/pg_task.so.

However, there are no pg_task files in /usr/share/postgresql/15/extension. 

This approach:
````
FROM postgres:latest
RUN apt-get update && apt-get --yes upgrade
RUN apt-get install --yes build-essential make postgresql-server-dev-15 git wget libkrb5-dev systemtap-sdt-dev
RUN git clone https://github.com/RekGRpth/pg_task.git
WORKDIR pg_task
RUN make
RUN make install
RUN rm -rf pg_task
````
still fails with `could not load library "/usr/lib/postgresql/15/lib/pg_task.so": /usr/lib/postgresql/15/lib/pg_task.so: undefined symbol: postgresql_query__plan__start_semaphore`

**Comment by RekGRpth** [comment]
> pg_task is not listed as an available extension

and it should not!

**Comment by RekGRpth** [comment]
> there are no pg_task files in /usr/share/postgresql/15/extension

and it should not!

**Comment by RekGRpth** [comment]
> startup fails

with which error?

**Comment by RekGRpth** [comment]
> still fails with `could not load library "/usr/lib/postgresql/15/lib/pg_task.so": /usr/lib/postgresql/15/lib/pg_task.so: undefined symbol: postgresql_query__plan__start_semaphore`

could You show build log?

**Comment by RekGRpth** [comment]
setting may be seen by
```sql
select * from pg_settings where name ~ 'pg_conf' or name ~ 'pg_work' or name ~ 'pg_task'
```

**Comment by RichMacDonald** [comment]
Both approaches compile and startup successfully. The `postgresql_query__plan__start_semaphore` went away when I tried today. One difference possibly is that I had a docker caching issue. I compiled with `--no-cache' this time.

pg_task does not appear in the list of available or installed extensions, however, the task table is present and the extension works.

Thank you very much.
